src/qibolab/transpilers/placer.py

```diff
@@ -0,0 +1,229 @@
+        else:
+            raise_error(TypeError, "Use dict or list to define mapping.")
+        if circuit is not None:
+            assert_placement(circuit, self.map)
+        else:
+            assert_mapping_consistency(self.map)
+        return self.map
+
+
+class Subgraph(Placer):
+    """
+    Subgraph isomorphism qubit placer, NP-complete it can take a long time
+    for large circuits. This initialization method may fail for very short circuits.
+
+    Attributes:
+        connectivity (networkx.Graph): chip connectivity.
+    """
+
+    def __init__(self, connectivity):
+        self.connectivity = connectivity
+
+    def __call__(self, circuit: Circuit):
+        """Find the initial layout of the given circuit using subgraph isomorphism.
+
+        Args:
+            circuit (qibo.models.Circuit): Circuit to be transpiled.
+        """
+        circuit_repr = create_circuit_repr(circuit)
+        if len(circuit_repr) < 3:
+            raise_error(
+                ValueError, "Circuit must contain at least two two qubit gates to implement subgraph placement."
+            )
+        circuit_subgraph = nx.Graph()
+        circuit_subgraph.add_nodes_from(range(self.connectivity.number_of_nodes()))
+        matcher = nx.algorithms.isomorphism.GraphMatcher(self.connectivity, circuit_subgraph)
+        i = 0
+        circuit_subgraph.add_edge(circuit_repr[i][0], circuit_repr[i][1])
+        while matcher.subgraph_is_monomorphic() == True:
+            result = matcher
+            i += 1
+            circuit_subgraph.add_edge(circuit_repr[i][0], circuit_repr[i][1])
+            matcher = nx.algorithms.isomorphism.GraphMatcher(self.connectivity, circuit_subgraph)
+            if self.connectivity.number_of_edges() == circuit_subgraph.number_of_edges() or i == len(circuit_repr) - 1:
+                keys = list(result.mapping.keys())
+                keys.sort()
+                return {i: result.mapping[i] for i in keys}
+        keys = list(result.mapping.keys())
+        keys.sort()
+        return {i: result.mapping[i] for i in keys}
+
+
+class Random(Placer):
+    """
+    Random initialization with greedy policy, let a maximum number of 2-qubit
+    gates can be applied without introducing any SWAP gate.
+
+    Attributes:
+        connectivity (networkx.Graph): chip connectivity.
+        samples (int): number of initial random layouts tested.
+    """
+
+    def __init__(self, connectivity, samples=100):
+        self.connectivity = connectivity
+        self.samples = samples
+
+    def __call__(self, circuit):
+        """Find an initial layout of the given circuit using random greedy algorithm.
+
+        Args:
+            circuit (qibo.models.Circuit): Circuit to be transpiled.
+        """
+        circuit_repr = create_circuit_repr(circuit)
+        nodes = self.connectivity.number_of_nodes()
+        keys = list(self.connectivity.nodes())
+        final_mapping = dict(zip(keys, list(range(nodes))))
```

Comment:
```suggestion
        final_mapping = dict(zip(keys, range(nodes)))
```

`zip` consumes iterables (as also `dict` and other functions in Python), so you don't need to iterate them ahead of times to make them lists.

I'm sure they are iterable, because even `list()` just consume iterables.

It is a different story for `keys` above, because you want to reuse, but an iterable can be iterated only once, in general.
However, I'm sure that here you're not iterated more than once, because `range()` is constructed inside `list()`, so you could not reference it elsewhere.